numba/core/dispatcher.py

```diff
@@ -969,6 +969,24 @@ def folded(args, kws):
                                                                       kws)[1]
                         raise e.bind_fold_arguments(folded)
                     self.add_overload(cres)
+
+                # Add additional overload for optional args
+                # For int8(optional(int8), optional(int8)) we should compile
+                # int8(opitonal(int8), omitted) and int8( omitted, omitted)
+                for i, arg in enumerate(reversed(args), 1):
+                    if not isinstance(arg, types.Optional):
+                        break
+                    args_variant = args[:-i] + (types.Omitted(None), ) * i
+                    if self.overloads.get(tuple(args_variant)) is not None:
+                        continue
+                    try:
+                        cres_variant = self._compiler.compile(args_variant,
+                                                              return_type)
+                    except Exception:
+                        pass
+                    else:
+                        self.add_overload(cres_variant)
+
```

Comment:
I don't think adding overloads here is the right approach. The problem in https://github.com/numba/numba/issues/7903 needs to be addressed at runtime---the array indexing implementation needs to check whether the index is an optional.

Reply:
Not the right approach - exactly.  I thought you will not read "draft" PR's so thank you for reading this.  
I've provided additional description above. First commit in PR solves index trouble, but there is one more issue. By changes in dispatcher i've tried to deal with OmittedArg, and yes that was wrong direction (but half of tests is green :) )
 
so I still diving in code and figuring out OmittedArg issue.